File: concrete/src/main/java/pcb/uwu/amount/quantity/Celsius.java
package pcb.uwu.amount.quantity;

import pcb.uwu.core.BigDecimalAmount;
import pcb.uwu.core.Magnitude;
import pcb.uwu.core.UnitAmount;
import pcb.uwu.units.quantity.TemperatureUnit;

import java.math.BigDecimal;
import java.math.MathContext;

import static pcb.uwu.unit.quantity.CelsiusUnit.CELSIUS;

public class Celsius extends Temperature {

	// region constructors

	public Celsius(Number value) {
		super(value, CELSIUS);
	}

	public Celsius(Number value, Magnitude magnitude) {
		super(value, magnitude, CELSIUS);
	}

	public Celsius(String value) {
		super(value, CELSIUS);
	}

	public Celsius(String value, Magnitude magnitude) {
		super(value, magnitude, CELSIUS);
	}

	public Celsius(BigDecimal value) {
		super(value, CELSIUS);
	}

	public Celsius(BigDecimal value, Magnitude magnitude) {
		super(value, magnitude, CELSIUS);
	}

	public Celsius(BigDecimalAmount amount) {
		super(amount, CELSIUS);
	}

	public Celsius(BigDecimalAmount amount, Magnitude magnitude) {
		super(amount, magnitude, CELSIUS);
	}

	// endregion

	// region implement UnitAmount

	@Override
	public Celsius plus(UnitAmount<TemperatureUnit> other, MathContext mathContext) {
		return new Celsius(plusAmount(other, mathContext));
	}

	@Override
	public Celsius minus(UnitAmount<TemperatureUnit> other, MathContext mathContext) {
		return new Celsius(minusAmount(other, mathContext));
	}

	@Override
	public Celsius multipliedBy(BigDecimal other, MathContext mathContext) {
		return new Celsius(multipliedByScalar(other, mathContext));
	}

	@Override
	public Celsius dividedBy(BigDecimal other, MathContext mathContext) {
		return new Celsius(dividedByScalar(other, mathContext));
	}

	// endregion
}

File: implementation/src/main/java/pcb/uwu/amount/base/Celsius.kt
package pcb.uwu.amount.base

import pcb.uwu.core.Magnitude
import pcb.uwu.core.Magnitude.NATURAL
import pcb.uwu.core.UnitAmount
import pcb.uwu.unit.base.CelsiusUnit
import pcb.uwu.unit.base.TemperatureUnit

class Celsius : Temperature
{
    @JvmOverloads
    constructor(amount: Number,
                magnitude: Magnitude = NATURAL)
            : super(amount, magnitude, CelsiusUnit)

    @JvmOverloads
    constructor(amount: String,
                magnitude: Magnitude = NATURAL)
            : super(amount, magnitude, CelsiusUnit)

    // region UnitAmount

    override fun plus(temperature: UnitAmount<TemperatureUnit>) =
        Celsius(this.amount + (temperature to CelsiusUnit).amount)

    override fun minus(temperature: UnitAmount<TemperatureUnit>) =
        Celsius(this.amount + (temperature to CelsiusUnit).amount)

    override fun times(scalar: Number) =
        Celsius(this.amount * scalar)

    override fun div(scalar: Number) =
        Celsius(this.amount / scalar)

    // endregion
}
